### PR TITLE
Add strategies to the lib

### PR DESCRIPTION
**Pull request by eraleis**

### CODE REVIEW

File: lib/fail_to_ban.rb [hunk 1]
```diff
@@ -1,61 +1,30 @@
...
-    :ok
+  def attempt
+    return :ok if @strategy.attempt
+    :blocked
   end
 
   def blocked?
-    return false if retry_count < PERMIT_FAIL_ATTEMPS || Time.now.to_i >= unlock_at
-    true
+    @strategy.blocked?
```
Comment (tlarrieu) [inline lib/fail_to_ban.rb]: [Forwardable](http://ruby-doc.org/stdlib-2.0.0/libdoc/forwardable/rdoc/Forwardable.html)

File: lib/fail_to_ban.rb [hunk 2]
```diff
@@ -1,61 +1,30 @@
...
   end
 
   def reset
-    @storage.del(id)
+    @strategy.reset
```
Comment (tlarrieu) [inline lib/fail_to_ban.rb]: [Forwardable](http://ruby-doc.org/stdlib-2.0.0/libdoc/forwardable/rdoc/Forwardable.html)

File: lib/fail_to_ban.rb [hunk 3]
```diff
@@ -1,61 +1,30 @@
...
     :ok
   end
 
   def unlock_at
-    @storage.hget(id, 'unlock_at').to_i
+    @strategy.unlock_at
```
Comment (tlarrieu) [inline lib/fail_to_ban.rb]: [Forwardable](http://ruby-doc.org/stdlib-2.0.0/libdoc/forwardable/rdoc/Forwardable.html)

File: lib/strategies/backoff_strategy.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    end
+
+    def attempt
+      return false if blocked?
+      increment_failed_attempts
+      true
+    end
+
+    def blocked?
+      return false if retry_count < @config[:permitted_attempts] || Time.now.to_i >= unlock_at
```
Comment (tlarrieu) [inline lib/strategies/backoff_strategy.rb]: when returning false if a condition and then true makes me feel we should use the condition itself as the return value

File: spec/fail_to_ban_spec.rb [hunk 5]
```diff
@@ -42,111 +42,112 @@ def del(k)
...
+          strategy: Strategies::BackoffStrategy
+        )
+      end
 
-    let(:key) { 'dev@jobteaser.com' }
+      let(:key) { 'dev@jobteaser.com' }
 
-    context 'when key is not blocked' do
+      context 'when key is not blocked' do
+        let(:data) do
```
Comment (tlarrieu) [inline spec/fail_to_ban_spec.rb]: this could easily be a one-liner

File: spec/fail_to_ban_spec.rb [hunk 6]
```diff
@@ -42,111 +42,112 @@ def del(k)
...
       end
 
-    end
-
-    context 'when retry > 3' do
-      let(:data) do
-        {}
-      end
+      context 'when retry exceeds the permitted_attempts' do
+        let(:data) do
```
Comment (tlarrieu) [inline spec/fail_to_ban_spec.rb]: this could easily be a one-liner

File: spec/fail_to_ban_spec.rb [hunk 7]
```diff
@@ -42,111 +42,112 @@ def del(k)
...
   let(:storage) { NullStorage.new(data) }
 
-  describe '#protect' do
-
-    subject { described_class.new(storage: storage, unique_key: key).protect }
-
-    let(:key) { 'dev@jobteaser.com' }
-
-    context 'when retry < 3' do
+  describe 'BackoffStrategy' do
```
Comment (tlarrieu) [inline spec/fail_to_ban_spec.rb]: I really think that we should still test the "core" class. If you do not see what to test or how, I'd be glad to setup a quick pairing ;)

File: lib/fail_to_ban/strategies/backoff_strategy.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        @storage.del(id)
+      end
+
+      def unlock_at
+        @storage.hget(id, 'unlock_at').to_i
+      end
+
+      private
+
+      attr_reader(:id)
```
Comment (gearnode) [inline lib/fail_to_ban/strategies/backoff_strategy.rb]: Anti-pattern

**Pull request merged**